### PR TITLE
Cleave tracker_scalar params to tracker_param

### PR DESCRIPTION
**Pull request by mgrauer**
This is ready for review.

The PR splits out params from tracker_scalar to a first class model.  It does not drop params from tracker_scalar though, as we'll need to do that in a separate PR so that we can migrate some live data in between this PR being deployed and that step of dropping params.

### CODE REVIEW

**Comment by jamiesnape** [comment]
Please rebase on the current master to fix StyleCI.

**Comment by mgrauer** [comment]
Todo:  update GUI where params are displayed

**Comment by mgrauer** [comment]
Ready for review.  Field tested today.

@jamiesnape @cpatrick

File: modules/tracker/controllers/components/ApiscalarComponent.php [hunk 1]
```diff
@@ -234,6 +235,19 @@ public function put($args)
         $scalarDao->setScalarId($scalarId);
         $scalarModel->save($scalarDao);
 
+        if (isset($params) && is_string($params)) {
+            $params = json_decode($params);
+            $paramModel = MidasLoader::loadModel('Param', $this->moduleName);
+            foreach ($params as $paramName => $paramValue) {
+                //* @var Tracker_ParamDao $paramDao * /
```
Comment (jamiesnape) [inline modules/tracker/controllers/components/ApiscalarComponent.php]: `/** .... */`

File: modules/tracker/database/upgrade/1.2.2.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+=========================================================================*/
+
+/**
+ * Upgrade the tracker module to version 1.2.2.
+ */
+class Tracker_Upgrade_1_2_2 extends MIDASUpgrade
```
Comment (jamiesnape) [inline modules/tracker/database/upgrade/1.2.2.php]: `/** Upgrade ... 1.2.2. */`

File: modules/tracker/models/dao/ParamDao.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            $this->setParamType('text');
+            $this->setTextValue($paramValue);
+        }
+    }
+
+    /**
+     * Get the value of the param, regardless of its type, returning either a
+     * numeric or a string.
+     *
+     * @return mixed value of the param
```
Comment (jamiesnape) [inline modules/tracker/models/dao/ParamDao.php]: `@return float|int|string` (and `null` if that can be returned).

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 4]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
             $curOutput = $outputs[$scalarDao->getTrend()->getMetricName()];
+            $metricToScalar[$scalarDao->getTrend()->getMetricName()] = $scalarDao;
             $this->assertEquals($curOutput->value, $scalarDao->getValue());
             $this->assertEquals($submissionDao->getKey(), $scalarDao->getSubmissionId());
         }
+
+        // Params should be a zero element array here.
+        $this->assertTrue(!($metricToScalar['metric_0']->getParams()));
+
+        $metric_1_params = $metricToScalar['metric_1']->getParams();
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camelcase

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 5]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
+        $metric_1_paramChecks = array(
+            'num_param' => array('found' => false, 'type' => 'numeric', 'val' => 19.0),
+            'text_param' => array('found' => false, 'type' => 'text', 'val' => 'metric1 text'),
+            'null_param' => array('found' => false, 'type' => 'text', 'val' => ''),
+        );
+
+        // Test that the params are saved as the correct type and value.
+        foreach ($metric_1_params as $param) {
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camelcase

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 6]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
+            if ($checks['type'] === 'numeric') {
+                $this->assertEquals($checks['val'], $param->getNumericValue());
+            } else {
+                $this->assertEquals($checks['val'], $param->getTextValue());
+            }
+            $metric_1_paramChecks[$param->getParamName()]['found'] = true;
+        }
+
+        // Test that all params are tested.
+        foreach ($metric_1_paramChecks as $checks) {
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camel case

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 7]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
+            $this->assertTrue($checks['found']);
+        }
+
+        $metric_2_params = $metricToScalar['metric_2']->getParams();
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camelcase

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 8]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
+        $metric_2_paramChecks = array(
+            'num_param' => array('found' => false, 'type' => 'numeric', 'val' => 20.0),
+            'text_param' => array('found' => false, 'type' => 'text', 'val' => 'metric2 text'),
+            'null_param' => array('found' => false, 'type' => 'text', 'val' => ''),
+        );
+
+        // Test that the params are saved as the correct type and value.
+        foreach ($metric_2_params as $param) {
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camelcase

File: modules/tracker/tests/controllers/ApiComponentTest.php [hunk 9]
```diff
@@ -60,12 +62,67 @@ public function testUploadScalarWithSubmission()
...
+            if ($checks['type'] === 'numeric') {
+                $this->assertEquals($checks['val'], $param->getNumericValue());
+            } else {
+                $this->assertEquals($checks['val'], $param->getTextValue());
+            }
+            $metric_2_paramChecks[$param->getParamName()]['found'] = true;
+        }
+
+        // Test that all params are tested.
+        foreach ($metric_2_paramChecks as $checks) {
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiComponentTest.php]: camelcase

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        $producerModel = MidasLoader::loadModel('Producer', 'tracker');
+        $producerDao = $producerModel->load(100);
+
+        $submissionModel = MidasLoader::loadModel('Submission', 'tracker');
+        $submissionDao = $submissionModel->getOrCreateSubmission($producerDao, $submissionUuid);
+        $submissionId = $submissionDao->getKey();
+
+        // Create two scalars, each tied to a separate trend.
+
+        $trendModel = MidasLoader::loadModel('Trend', 'tracker');
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: /** @var ..*/

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+        $producerModel = MidasLoader::loadModel('Producer', 'tracker');
+        $producerDao = $producerModel->load(100);
+
+        $submissionModel = MidasLoader::loadModel('Submission', 'tracker');
+        $submissionDao = $submissionModel->getOrCreateSubmission($producerDao, $submissionUuid);
+        $submissionId = $submissionDao->getKey();
+
+        // Create two scalars, each tied to a separate trend.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: `/** @var ... */`

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        foreach ($paramChecks as $checks) {
+            $this->assertTrue($checks['found']);
+        }
+
+        $this->assertEquals($scalarDao0->getValue(), $scalarValue0);
+
+        // Scalar the second.
+
+        $metric1trend = $trendModel->load(2);
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: camelcase

File: modules/tracker/tests/controllers/ApiScalarComponentTest.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @throws Zend_Exception
+     */
+    public function testPUT()
+    {
+        $usersFile = $this->loadData('User', 'default');
+        $userDao = $this->User->load($usersFile[0]->getKey());
+
+        // Create a scalar attached to a trend.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiScalarComponentTest.php]: `/** @var ... */`

File: modules/tracker/tests/controllers/ApiScalarComponentTest.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        $trendModel = MidasLoader::loadModel('Trend', 'tracker');
+        $trend = $trendModel->load(1);
+
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiScalarComponentTest.php]: `/** @var ... */`

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        foreach ($paramChecks as $checks) {
+            $this->assertTrue($checks['found']);
+        }
+
+        $this->assertEquals($scalarDao1->getValue(), $scalarValue1);
+
+        // Delete scalars and ensure params are deleted.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: delete newline

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+            $scalarValue1,
+            $userDao,
+            true,
+            true,
+            'http://buildresultsurl',
+            'dev',
+            $params1);
+
+        // Ensure value and params are properly set on scalar.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: delete newline

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        foreach ($paramChecks as $checks) {
+            $this->assertTrue($checks['found']);
+        }
+
+        $this->assertEquals($scalarDao0->getValue(), $scalarValue0);
+
+        // Scalar the second.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: delete newline

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            $scalarValue0,
+            $userDao,
+            true,
+            true,
+            'http://buildresultsurl',
+            'master',
+            $params0);
+
+        // Ensure value and params are properly set on scalar.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: delete newline

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+        $userDao = $userModel->load($usersFile[0]->getKey());
+
+        $producerModel = MidasLoader::loadModel('Producer', 'tracker');
+        $producerDao = $producerModel->load(100);
+
+        $submissionModel = MidasLoader::loadModel('Submission', 'tracker');
+        $submissionDao = $submissionModel->getOrCreateSubmission($producerDao, $submissionUuid);
+        $submissionId = $submissionDao->getKey();
+
+        // Create two scalars, each tied to a separate trend.
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: delete newline

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+=========================================================================*/
+
+/** test ScalarModel */
+class Tracker_ScalarModelTest extends DatabaseTestCase
+{
+    /** set up tests */
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: `Set up tests.`

File: modules/tracker/tests/models/base/ScalarModelTest.php [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+         http://www.apache.org/licenses/LICENSE-2.0.txt
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+=========================================================================*/
+
+/** test ScalarModel */
```
Comment (jamiesnape) [inline modules/tracker/tests/models/base/ScalarModelTest.php]: `Test the scalar model.`

File: modules/tracker/tests/controllers/ApiScalarComponentTest.php [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+     * Test updating an existing scalar with a set of params, via PUT.
+     *
+     * @throws Zend_Exception
+     */
+    public function testPUT()
+    {
+        $usersFile = $this->loadData('User', 'default');
+        $userDao = $this->User->load($usersFile[0]->getKey());
+
+        // Create a scalar attached to a trend.
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiScalarComponentTest.php]: delete newline

File: modules/tracker/tests/controllers/ApiScalarComponentTest.php [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            'params' => json_encode($params),
+            'token' => $token,
+        );
+
+        $this->resetAll();
+        $this->params = $restParams;
+        $resp = $this->_callRestApi('PUT', '/tracker/scalar/'.$scalar->getScalarId());
+
+        // Ensure params are properly set on scalar.
+
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiScalarComponentTest.php]: delete newline

File: modules/tracker/models/dao/ParamDao.php [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public function getParamValue()
+    {
+        if ($this->getParamType() === 'numeric') {
+            return $this->getNumericValue();
+        } else {
+            return $this->getTextValue();
+        }
```
Comment (jamiesnape) [inline modules/tracker/models/dao/ParamDao.php]:
```php
if ($this->getParamType() === 'numeric') {
    return $this->getNumericValue();
}

return $this->getTextValue();
```

File: modules/tracker/database/upgrade/1.2.2.php [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+        $this->db->query(
+            'CREATE TABLE IF NOT EXISTS "tracker_param" ('.
+            '    "param_id" serial PRIMARY KEY,'.
+            '    "scalar_id" bigint NOT NULL,'.
+            '    "param_name" character varying(255) NOT NULL,'.
+            '    "param_type" text CHECK (param_type in (\'text\', \'numeric\')),'.
+            '    "text_value" text,'.
+            '    "numeric_value" double precision);'
+        );
+        $this->db->query('CREATE INDEX "tracker_param_param_name_idx" on "tracker_param" ("param_name");');
```
Comment (jamiesnape) [inline modules/tracker/database/upgrade/1.2.2.php]: `ON`

File: modules/tracker/database/upgrade/1.2.2.php [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Upgrade the tracker module to version 1.2.2.
+ */
+class Tracker_Upgrade_1_2_2 extends MIDASUpgrade
+{
+    /** Upgrade a MySQL database. */
+    public function mysql()
+    {
+        $this->db->query(
+            'CREATE TABLE IF NOT EXISTS `tracker_param` ('.
```
Comment (jamiesnape) [inline modules/tracker/database/upgrade/1.2.2.php]: Use double quotes so you do not need to escape.

File: modules/tracker/database/upgrade/1.2.2.php [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+            '    PRIMARY KEY (`param_id`),'.
+            '    KEY (`param_name`)'.
+            ') DEFAULT CHARSET=utf8;');
+    }
+
+    /** Upgrade a PostgreSQL database. */
+    public function pgsql()
+    {
+        $this->db->query(
+            'CREATE TABLE IF NOT EXISTS "tracker_param" ('.
```
Comment (jamiesnape) [inline modules/tracker/database/upgrade/1.2.2.php]: PostgreSQL queries in upgrade scripts typically omit the double quotes.

**Comment by mgrauer** [comment]
QA Feedback addressed.

**Pull request merged**